pkg/apis/volumesnapshot/v1beta1/types.go

```diff
@@ -0,0 +1,354 @@
+	metav1.TypeMeta `json:",inline"`
+	// Standard object's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	// spec defines the desired characteristics of a snapshot requested by a user.
+	// More info: https://kubernetes.io/docs/concepts/storage/volume-snapshots#volumesnapshots
+	// Required.
+	Spec VolumeSnapshotSpec `json:"spec" protobuf:"bytes,2,opt,name=spec"`
+
+	// status represents the current information of a snapshot.
+	// NOTE: status can be modified by sources other than system controllers,
+	// and must not be depended upon for accuracy.
+	// Controllers should only use information from the VolumeSnapshotContent object
+	// after verifying that the binding is accurate and complete.
+	// +optional
+	Status VolumeSnapshotStatus `json:"status,omitempty" protobuf:"bytes,3,opt,name=status"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+// VolumeSnapshotList is a list of VolumeSnapshot objects
+type VolumeSnapshotList struct {
+	metav1.TypeMeta `json:",inline"`
+	// +optional
+	metav1.ListMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	// List of VolumeSnapshots
+	Items []VolumeSnapshot `json:"items" protobuf:"bytes,2,rep,name=items"`
+}
+
+// VolumeSnapshotSpec describes the common attributes of a volume snapshot.
+type VolumeSnapshotSpec struct {
+	// source specifies where a snapshot will be created from.
+	// Required.
+	Source VolumeSnapshotSource `json:"source" protobuf:"bytes,1,opt,name=source"`
+
+	// volumeSnapshotClassName is the name of the VolumeSnapshotClass requested by the VolumeSnapshot.
+	// If not specified, the default snapshot class will be used if one exists.
+	// If not specified, and there is no default snapshot class, dynamic snapshot creation will fail.
+	// More info: https://kubernetes.io/docs/concepts/storage/volume-snapshot-classes
+	// +optional
+	VolumeSnapshotClassName *string `json:"volumeSnapshotClassName,omitempty" protobuf:"bytes,2,opt,name=volumeSnapshotClassName"`
+}
+
+// VolumeSnapshotSource represents the source of a snapshot.
+// Exactly one of its members must be set.
+// Members in VolumeSnapshotSource are immutable.
+// TODO(xiangqian): Add a webhook to ensure that VolumeSnapshotSource members
+// will not be updated once specified.
+type VolumeSnapshotSource struct {
+	// persistentVolumeClaimName specifies the name of the PersistentVolumeClaim
+	// object in the same namespace as the VolumeSnapshot object where the
+	// snapshot should be dynamically taken from.
+	// +optional
+	PersistentVolumeClaimName *string `json:"persistentVolumeClaimName,omitempty" protobuf:"bytes,1,opt,name=persistentVolumeClaimName"`
+
+	// volumeSnapshotContentName specifies the name of a pre-existing VolumeSnapshotContent
+	// object a user asks to statically bind the VolumeSnapshot object to.
```

Comment:
I want to avoid using the term "static binding". I think we can remove this part of the sentence and just keep the first half.

Reply:
done